vcr/oidc4vci/service/identifiers.go

```diff
@@ -0,0 +1,164 @@
+ *
+ */
+
+package service
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/vcr/log"
+	"github.com/nuts-foundation/nuts-node/vcr/oidc4vci"
+	"github.com/nuts-foundation/nuts-node/vdr/didservice"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+)
+
+// IdentifierResolver defines the interface for resolving OpenID4VCI identifiers (of wallet and issuer).
+// The identifier is the base URL of the issuer or wallet, at which well-known endpoints can be found.
+type IdentifierResolver interface {
+	Resolve(id did.DID) (string, error)
+}
+
+var _ IdentifierResolver = DIDIdentifierResolver{}
+
+// DIDIdentifierResolver is a IdentifierResolver that uses resolves identifiers from DID documents.
+type DIDIdentifierResolver struct {
+	ServiceResolver didservice.ServiceResolver
+}
+
+func (i DIDIdentifierResolver) Resolve(id did.DID) (string, error) {
+	service, err := i.ServiceResolver.Resolve(didservice.MakeServiceReference(id, types.BaseURLServiceType), didservice.DefaultMaxServiceReferenceDepth)
+	if didservice.IsFunctionalResolveError(err) {
+		return "", nil
+	} else if err != nil {
+		return "", fmt.Errorf("unable to resolve %s service: %w", types.BaseURLServiceType, err)
+	}
+	var result string
+	_ = service.UnmarshalServiceEndpoint(&result)
+	if result != "" {
+		result = core.JoinURLPaths(result, "n2n", "identity", url.PathEscape(id.String()))
+	}
+	return result, nil
+}
+
+// NewTLSIdentifierResolver creates a IdentifierResolver that tries to derive the identifier from the TLS certificate if it can't be resolved using the DID document.
+// It does so by constructing the identifier from the CommonName and SubjectAlternativeNames of the certificate and requesting metadata.
+func NewTLSIdentifierResolver(underlying IdentifierResolver, config *tls.Config) IdentifierResolver {
+	result := tlsIdentifierResolver{
+		underlying:       underlying,
+		config:           config,
+		cachedIdentifier: new(atomic.Pointer[string]),
+		lastAttempt:      new(atomic.Pointer[time.Time]),
+	}
+	result.lastAttempt.Store(new(time.Time))
+	return result
+}
+
+const tlsAttemptInterval = time.Minute
+
+var tlsIdentifierResolverPort = 443
+
+func SetTLSIdentifierResolverPort(port int) {
+	tlsIdentifierResolverPort = port
+}
+
+var _ IdentifierResolver = tlsIdentifierResolver{}
+
+type tlsIdentifierResolver struct {
+	underlying       IdentifierResolver
+	config           *tls.Config
+	cachedIdentifier *atomic.Pointer[string]
+	// lastAttempt is the time at which the last attempt to resolve the identifier from the TLS certificate was made.
+	// It is used to prevent spamming the local node, since it could be called on each OpenID4VCI request.
+	lastAttempt *atomic.Pointer[time.Time]
+}
+
+func (t tlsIdentifierResolver) Resolve(id did.DID) (string, error) {
```

Comment:
Weird caching. The primary resolve loads a DID Document from database which is a heavier operation that checking the certificate data which is already in-memory?